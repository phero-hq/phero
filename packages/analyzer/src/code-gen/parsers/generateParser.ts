import ts from "typescript"
import { Model } from "../../parseSamenApp"
import generateParserFromModel from "./../parsers/generateParserFromModel"
import generateParserModel from "./../parsers/generateParserModel"

const exportModifier = ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)
const staticModifier = ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)

export default function generateModelParser(
  model: Model,
  typeChecker: ts.TypeChecker,
): ts.ClassDeclaration {
  const rootParserModel = generateParserModel(typeChecker, model, "data")
  if (!rootParserModel.rootTypeParser) {
    // TODO enum
    throw new Error("Expected interface or typealias")
  }

  const parserStatement: ts.Statement = generateParserFromModel(rootParserModel)

  return ts.factory.createClassDeclaration(
    undefined,
    [exportModifier],
    `${rootParserModel.rootTypeParser.baseTypeName}Parser`,
    undefined,
    undefined,
    [
      ts.factory.createMethodDeclaration(
        undefined,
        [staticModifier],
        undefined,
        "parse",
        undefined,
        rootParserModel.rootTypeParser.typeParameters.map((p) =>
          ts.factory.createTypeParameterDeclaration(p.typeName),
        ),
        [
          ts.factory.createParameterDeclaration(
            undefined,
            undefined,
            undefined,
            "data",
            undefined,
            ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
            undefined,
          ),
          ...rootParserModel.rootTypeParser.typeParameters.map(
            (typeParam, position) =>
              ts.factory.createParameterDeclaration(
                undefined,
                undefined,
                undefined,
                ts.factory.createIdentifier(`t${position}`),
                typeParam.defaultParser
                  ? ts.factory.createToken(ts.SyntaxKind.QuestionToken)
                  : undefined,
                ts.factory.createFunctionTypeNode(
                  undefined,
                  [
                    ts.factory.createParameterDeclaration(
                      undefined,
                      undefined,
                      undefined,
                      ts.factory.createIdentifier("data"),
                      undefined,
                      ts.factory.createKeywordTypeNode(
                        ts.SyntaxKind.AnyKeyword,
                      ),
                      undefined,
                    ),
                  ],
                  ts.factory.createTypeReferenceNode(
                    ts.factory.createIdentifier("ParseResult"),
                    [
                      ts.factory.createTypeReferenceNode(
                        ts.factory.createIdentifier(typeParam.typeName),
                        undefined,
                      ),
                    ],
                  ),
                ),
                undefined,
              ),
          ),
        ],
        ts.factory.createTypeReferenceNode(
          ts.factory.createIdentifier("ParseResult"),
          [
            ts.factory.createTypeReferenceNode(
              rootParserModel.rootTypeParser.typeName,
              undefined,
            ),
          ],
        ),
        generateParserBody(
          rootParserModel.rootTypeParser.typeName,
          parserStatement,
        ),
      ),
    ],
  )
}

export function generateParserBody(
  returnTypeString: string,
  parserStatement: ts.Statement,
): ts.Block {
  return ts.factory.createBlock(
    [
      ts.factory.createVariableStatement(
        undefined,
        ts.factory.createVariableDeclarationList(
          [
            ts.factory.createVariableDeclaration(
              ts.factory.createIdentifier("errors"),
              undefined,
              ts.factory.createArrayTypeNode(
                ts.factory.createTypeReferenceNode(
                  ts.factory.createIdentifier("ValidationError"),
                  undefined,
                ),
              ),
              ts.factory.createArrayLiteralExpression([], false),
            ),
          ],
          ts.NodeFlags.Const,
        ),
      ),
      ts.factory.createVariableStatement(
        undefined,
        ts.factory.createVariableDeclarationList(
          [
            ts.factory.createVariableDeclaration(
              ts.factory.createIdentifier("result"),
              undefined,
              ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
              undefined,
            ),
          ],
          ts.NodeFlags.Let,
        ),
      ),

      parserStatement,

      ts.factory.createIfStatement(
        ts.factory.createPropertyAccessExpression(
          ts.factory.createIdentifier("errors"),
          ts.factory.createIdentifier("length"),
        ),
        ts.factory.createBlock(
          [
            ts.factory.createReturnStatement(
              ts.factory.createObjectLiteralExpression(
                [
                  ts.factory.createPropertyAssignment(
                    ts.factory.createIdentifier("ok"),
                    ts.factory.createFalse(),
                  ),
                  ts.factory.createShorthandPropertyAssignment(
                    ts.factory.createIdentifier("errors"),
                    undefined,
                  ),
                ],
                true,
              ),
            ),
          ],
          true,
        ),
        undefined,
      ),
      ts.factory.createReturnStatement(
        ts.factory.createObjectLiteralExpression(
          [
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier("ok"),
              ts.factory.createTrue(),
            ),
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier("result"),
              ts.factory.createAsExpression(
                ts.factory.createIdentifier("result"),
                ts.factory.createTypeReferenceNode(returnTypeString, undefined),
              ),
            ),
          ],
          true,
        ),
      ),
    ],
    true,
  )
}
