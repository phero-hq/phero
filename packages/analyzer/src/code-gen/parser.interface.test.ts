import ts from "typescript"
import { compileStatement, printCode } from "../tsTestUtils"
import generateParser from "./parsers/generateParser"

describe("Parsers", () => {
  describe("for an interface", () => {
    test("with no members", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("number member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: number
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("string member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: string
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("boolean member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: boolean
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("null member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: null
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("optional member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a?: string
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("string literal member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: "aap"
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("number literal member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: 123
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("true literal member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: true
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)

      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("string array member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: string[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("optional string array member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a?: string[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("number array member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: number[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("boolean array member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: boolean[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("object literal member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: {
              b: number
            }
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("object literal array member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: {
              b: number
            }[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("optional object literal member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a?: {
              b: number
            }
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("optional object literal with optional member", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a?: {
              b?: number
            }
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
    test("deep nested object", () => {
      const { statement: model, typeChecker } = compileStatement(
        `
          interface MyModel {
            a: {
              b?: {
                c: {
                  d: {
                    e: {
                      f: boolean[]
                    }
                  }[]
                }
              }
            }[]
          }
        `,
        ts.SyntaxKind.InterfaceDeclaration,
      )

      const parserDeclaration = generateParser(model, typeChecker)
      expect(printCode(parserDeclaration)).toMatchSnapshot()
    })
  })
})
