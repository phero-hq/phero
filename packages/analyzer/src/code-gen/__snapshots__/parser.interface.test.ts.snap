// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsers for an interface boolean array member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (!Array.isArray(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not an array\`
                });
            }
            else {
                result[\\"a\\"] = [];
                for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                    if (typeof data[\\"a\\"][it_0] !== \\"boolean\\") {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`not a boolean\`
                        });
                    }
                    else
                        result[\\"a\\"][it_0] = data[\\"a\\"][it_0];
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface boolean member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (typeof data[\\"a\\"] !== \\"boolean\\") {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not a boolean\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface deep nested object 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (!Array.isArray(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not an array\`
                });
            }
            else {
                result[\\"a\\"] = [];
                for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                    if (typeof data[\\"a\\"][it_0] !== \\"object\\" || data[\\"a\\"][it_0] === null) {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`null or not an object\`
                        });
                    }
                    else {
                        result[\\"a\\"][it_0] = {};
                        if (data[\\"a\\"][it_0][\\"b\\"] !== undefined)
                            if (typeof data[\\"a\\"][it_0][\\"b\\"] !== \\"object\\" || data[\\"a\\"][it_0][\\"b\\"] === null) {
                                errors.push({
                                    path: \`data.a[\${it_0}].b\`,
                                    message: \`null or not an object\`
                                });
                            }
                            else {
                                result[\\"a\\"][it_0][\\"b\\"] = {};
                                if (typeof data[\\"a\\"][it_0][\\"b\\"][\\"c\\"] !== \\"object\\" || data[\\"a\\"][it_0][\\"b\\"][\\"c\\"] === null) {
                                    errors.push({
                                        path: \`data.a[\${it_0}].b.c\`,
                                        message: \`null or not an object\`
                                    });
                                }
                                else {
                                    result[\\"a\\"][it_0][\\"b\\"][\\"c\\"] = {};
                                    if (!Array.isArray(data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"])) {
                                        errors.push({
                                            path: \`data.a[\${it_0}].b.c.d\`,
                                            message: \`not an array\`
                                        });
                                    }
                                    else {
                                        result[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"] = [];
                                        for (let it_1 = 0; it_1 < data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"].length; it_1++)
                                            if (typeof data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1] !== \\"object\\" || data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1] === null) {
                                                errors.push({
                                                    path: \`data.a[\${it_0}].b.c.d[\${it_1}]\`,
                                                    message: \`null or not an object\`
                                                });
                                            }
                                            else {
                                                result[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1] = {};
                                                if (typeof data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"] !== \\"object\\" || data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"] === null) {
                                                    errors.push({
                                                        path: \`data.a[\${it_0}].b.c.d[\${it_1}].e\`,
                                                        message: \`null or not an object\`
                                                    });
                                                }
                                                else {
                                                    result[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"] = {};
                                                    if (!Array.isArray(data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"])) {
                                                        errors.push({
                                                            path: \`data.a[\${it_0}].b.c.d[\${it_1}].e.f\`,
                                                            message: \`not an array\`
                                                        });
                                                    }
                                                    else {
                                                        result[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"] = [];
                                                        for (let it_2 = 0; it_2 < data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"].length; it_2++)
                                                            if (typeof data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"][it_2] !== \\"boolean\\") {
                                                                errors.push({
                                                                    path: \`data.a[\${it_0}].b.c.d[\${it_1}].e.f[\${it_2}]\`,
                                                                    message: \`not a boolean\`
                                                                });
                                                            }
                                                            else
                                                                result[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"][it_2] = data[\\"a\\"][it_0][\\"b\\"][\\"c\\"][\\"d\\"][it_1][\\"e\\"][\\"f\\"][it_2];
                                                    }
                                                }
                                            }
                                    }
                                }
                            }
                    }
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface null member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (typeof data[\\"a\\"] !== null) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not null\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface number array member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (!Array.isArray(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not an array\`
                });
            }
            else {
                result[\\"a\\"] = [];
                for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                    if (typeof data[\\"a\\"][it_0] !== \\"number\\") {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`not a number\`
                        });
                    }
                    else if (isNaN(data[\\"a\\"][it_0])) {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`invalid number\`
                        });
                    }
                    else
                        result[\\"a\\"][it_0] = data[\\"a\\"][it_0];
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface number literal member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== 123) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not 123\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface number member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (typeof data[\\"a\\"] !== \\"number\\") {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not a number\`
                });
            }
            else if (isNaN(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`invalid number\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface object literal array member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (!Array.isArray(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not an array\`
                });
            }
            else {
                result[\\"a\\"] = [];
                for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                    if (typeof data[\\"a\\"][it_0] !== \\"object\\" || data[\\"a\\"][it_0] === null) {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`null or not an object\`
                        });
                    }
                    else {
                        result[\\"a\\"][it_0] = {};
                        if (typeof data[\\"a\\"][it_0][\\"b\\"] !== \\"number\\") {
                            errors.push({
                                path: \`data.a[\${it_0}].b\`,
                                message: \`not a number\`
                            });
                        }
                        else if (isNaN(data[\\"a\\"][it_0][\\"b\\"])) {
                            errors.push({
                                path: \`data.a[\${it_0}].b\`,
                                message: \`invalid number\`
                            });
                        }
                        else
                            result[\\"a\\"][it_0][\\"b\\"] = data[\\"a\\"][it_0][\\"b\\"];
                    }
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface object literal member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (typeof data[\\"a\\"] !== \\"object\\" || data[\\"a\\"] === null) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`null or not an object\`
                });
            }
            else {
                result[\\"a\\"] = {};
                if (typeof data[\\"a\\"][\\"b\\"] !== \\"number\\") {
                    errors.push({
                        path: \\"data.a.b\\",
                        message: \`not a number\`
                    });
                }
                else if (isNaN(data[\\"a\\"][\\"b\\"])) {
                    errors.push({
                        path: \\"data.a.b\\",
                        message: \`invalid number\`
                    });
                }
                else
                    result[\\"a\\"][\\"b\\"] = data[\\"a\\"][\\"b\\"];
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface optional member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== undefined)
                if (typeof data[\\"a\\"] !== \\"string\\") {
                    errors.push({
                        path: \\"data.a\\",
                        message: \`not a string\`
                    });
                }
                else
                    result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface optional object literal member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== undefined)
                if (typeof data[\\"a\\"] !== \\"object\\" || data[\\"a\\"] === null) {
                    errors.push({
                        path: \\"data.a\\",
                        message: \`null or not an object\`
                    });
                }
                else {
                    result[\\"a\\"] = {};
                    if (typeof data[\\"a\\"][\\"b\\"] !== \\"number\\") {
                        errors.push({
                            path: \\"data.a.b\\",
                            message: \`not a number\`
                        });
                    }
                    else if (isNaN(data[\\"a\\"][\\"b\\"])) {
                        errors.push({
                            path: \\"data.a.b\\",
                            message: \`invalid number\`
                        });
                    }
                    else
                        result[\\"a\\"][\\"b\\"] = data[\\"a\\"][\\"b\\"];
                }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface optional object literal with optional member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== undefined)
                if (typeof data[\\"a\\"] !== \\"object\\" || data[\\"a\\"] === null) {
                    errors.push({
                        path: \\"data.a\\",
                        message: \`null or not an object\`
                    });
                }
                else {
                    result[\\"a\\"] = {};
                    if (data[\\"a\\"][\\"b\\"] !== undefined)
                        if (typeof data[\\"a\\"][\\"b\\"] !== \\"number\\") {
                            errors.push({
                                path: \\"data.a.b\\",
                                message: \`not a number\`
                            });
                        }
                        else if (isNaN(data[\\"a\\"][\\"b\\"])) {
                            errors.push({
                                path: \\"data.a.b\\",
                                message: \`invalid number\`
                            });
                        }
                        else
                            result[\\"a\\"][\\"b\\"] = data[\\"a\\"][\\"b\\"];
                }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface optional string array member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== undefined)
                if (!Array.isArray(data[\\"a\\"])) {
                    errors.push({
                        path: \\"data.a\\",
                        message: \`not an array\`
                    });
                }
                else {
                    result[\\"a\\"] = [];
                    for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                        if (typeof data[\\"a\\"][it_0] !== \\"string\\") {
                            errors.push({
                                path: \`data.a[\${it_0}]\`,
                                message: \`not a string\`
                            });
                        }
                        else
                            result[\\"a\\"][it_0] = data[\\"a\\"][it_0];
                }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface string array member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (!Array.isArray(data[\\"a\\"])) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not an array\`
                });
            }
            else {
                result[\\"a\\"] = [];
                for (let it_0 = 0; it_0 < data[\\"a\\"].length; it_0++)
                    if (typeof data[\\"a\\"][it_0] !== \\"string\\") {
                        errors.push({
                            path: \`data.a[\${it_0}]\`,
                            message: \`not a string\`
                        });
                    }
                    else
                        result[\\"a\\"][it_0] = data[\\"a\\"][it_0];
            }
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface string literal member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== \\"aap\\") {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not 'aap'\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface string member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (typeof data[\\"a\\"] !== \\"string\\") {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not a string\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface true literal member 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
            if (data[\\"a\\"] !== true) {
                errors.push({
                    path: \\"data.a\\",
                    message: \`not true\`
                });
            }
            else
                result[\\"a\\"] = data[\\"a\\"];
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;

exports[`Parsers for an interface with no members 1`] = `
"export class MyModelParser {
    static parse(data: any): ParseResult<MyModel> {
        const errors: ValidationError[] = [];
        let result: any;
        if (typeof data !== \\"object\\" || data === null) {
            errors.push({
                path: \\"data\\",
                message: \`null or not an object\`
            });
        }
        else {
            result = {};
        }
        if (errors.length) {
            return {
                ok: false,
                errors
            };
        }
        return {
            ok: true,
            result: result as MyModel
        };
    }
}"
`;
